adding a new getitem method

shilpiprd opened the pull request:
I figured that get_item method is most closely related to gradients.py file which is inside dipy/core/gradients.py folder because i saw the bvech bvals method. 
# Example usage
#since we can call a getItem method by simply using square brackets, 
gtab = MyGradientTable(gradients)
gtab_selected = gtab[2:10]  # Select b-values from 2 to 10. 

Please let me know if I need to make any changes to this method.

arokem commented:
Yeah - I like it! Could you please add a couple of tests of this new functionality?

shilpiprd commented:
> Yeah - I like it! Could you please add a couple of tests of this new functionality?

I've added the tests, please review them.

shilpiprd commented:
Sure

skoudoro commented on dipy/core/gradients.py:
```diff
@@ -210,7 +210,26 @@ def bvecs(self):
+                                           small_delta=0.01,
+                                           b0_threshold=50,
+                                           atol=1e-2, btens=None)
+    #gave default value for small_delta, big_delta and b0_threshold
```
Why default value?  I recommend to get the value already initialized like  `self.big_delta , self.small_delta , etc`

shilpiprd replied:
You rea right. I'll change this as well.

shilpiprd commented:
I updated the getitem method, and realised that the test file needed to be changed as well.I changed the test file in this branch. Should I close the #2750 now? 
Please review and let me know if any other changes needs to be made.

arokem commented:
Yeah. I'll close the other one.

arokem commented on dipy/core/gradients.py:
```diff
@@ -210,7 +210,25 @@ def bvecs(self):
         return self.gradients / denom
-
+    
+    def __getitem__(self, idx, threshold = 200):
```
I would set the threshold to be 0 per default. 

One more thought is to implement a lower and upper bound, so that if you index: 

```
gtab[:, 100:1000]
```

you would get a new gradient table with all the b-values between 100 and 1,000. I can see some use-cases for this kind of selection.

shilpiprd replied:
I was thinking that we can probably remove the threshold argument and associated code, as it is no longer needed with the new indexing scheme.
Since now we could have: def __getitem__(self, bmin=0, bmax=np.inf):
And to create mask we can simply use: mask = (self.bvals >= bmin) & (self.bvals <= bmax). Using this to create mask we can remove a lot of unnecessary code as well. 
and # Select b-values between 100 and 1000, 
gtab = gtab[bmin=100, bmax=1000]
What do you think of this?

skoudoro replied:
I like the idea! 

however, a deprecation cycle will be needed. a lot people are using this default behavior so we will have to keep both behavior for 1 year (2-3 releases) at least.

shilpiprd replied:
So are you sugggesting something like this? 
def __getitem__(self, bmin=0, bmax=np.inf, threshold=0):
    # Use the new behavior if bmin and bmax are provided
    if bmin != 0 or bmax != np.inf:
        mask = (self.bvals >= bmin) & (self.bvals <= bmax)
    else:
        # Use the old behavior if bmin and bmax are not provided
        mask = self.bvals > threshold

skoudoro replied:
> But I don't quite understand the part where the interp parameter is being deprecated from version 1.19 onwards. Can you explain more on that.

typo, replaced by `threshold`

shilpiprd replied:
Oh right, so should i create a commit for this? I mean the method looks correct.

skoudoro replied:
yes, update the code and test it to make sure it works as expected.

shilpiprd replied:
I modified the gradients file and tests_gradients.py as well. Please review it.

shilpiprd replied:
Could somebody please manually trigger the checks on my last commit?

arokem reviewed (commented):
I've thought about this a bit more. What do you think about indexing with a lower and upper bound on b-values?

shilpiprd commented:
The test file is passing all checks. It is failing one check but I can't seem to figure out where the error is because it says PASSED everywhere. 
Is there a need to look into this ? or is it mergable?

skoudoro commented on dipy/core/gradients.py:
```diff
@@ -210,7 +211,26 @@ def bvecs(self):
-
+    
+    @deprecated_params('threshold', None, since='1.17', until='1.19')
+    def getitem(self, bmin=0, bmax=np.inf, threshold=0):
```
I think, we got lost here.  

The goal was to have the slicing capabilities. However, we loose it from your initial implementation.  
we can not do anymore gtab[100:200].

we should get back to that. we just need a warning if the slicing is really different from the `b0_treshold`.

Sorry for that. Can you revert your code, use `__get_item__`, and in case of a slice, you check if the lower bound is similar to `b0_thresold`. If yes, nothing to do, if no, just update self.b0_thresold to this value.

Does it make sense?   we do not need(bmin, bmax, thresold)

shilpiprd replied:
Sure, you're right. We did get deviated from the original method. I changed it, Did you mean something like this? If you'd agree with this, i'll change the test cases as well.

skoudoro reviewed (commented):
Hi @shilpiprd,

See below my comment and sorry for the late review.

skoudoro commented on dipy/core/gradients.py:
```diff
@@ -210,7 +211,26 @@ def bvecs(self):
-
+    
+    def getitem(self, bmin=0, bmax=np.inf):
+        if bmin != self.b0_threshold:
+            self.b0_threshold = bmin
+            warn("Updating b0_threshold to {} for slicing.".format(bmin), UserWarning, stacklevel=2)
```
yes, I mean something like this

skoudoro commented on dipy/core/gradients.py:
```diff
@@ -210,7 +211,26 @@ def bvecs(self):
         return self.gradients / denom
-
+    
+    def getitem(self, bmin=0, bmax=np.inf):
```
here it should be __get_item__(self, idx):

shilpiprd replied:
Sure, I'll change this

skoudoro commented on dipy/core/gradients.py:
```diff
@@ -210,7 +211,26 @@ def bvecs(self):
+            self.b0_threshold = bmin
+            warn("Updating b0_threshold to {} for slicing.".format(bmin), UserWarning, stacklevel=2)
+        # priority to bmin in all other cases
+        mask = (self.bvals >= bmin) & (self.bvals <= bmax)
```
like you did before, check if `slice` or just `int`

shilpiprd replied:
I'm sorry I didn't understand. Can u elaborate on what needs to be checked?

shilpiprd replied:
Were you referring to bmin value?

shilpiprd replied:
Oh, sorry. I understood what you meant.

shilpiprd commented:
I've added the idx parameter and modified the test file as well. Can someone please review it and allow for workflow checks to run?

shilpiprd commented:
I noticed that the failing checks are not from the method that I"ve implemented.

skoudoro commented on dipy/core/tests/test_gradients.py:
```diff
@@ -461,6 +460,26 @@ def test_generate_bvecs():
+    gtab = GradientTable(gradients) 
+
+    # Test with a single index
+    gtab_slice1 = gtab.__get_item__(1)
```
the goal was to do `gtab[1]` for example

skoudoro commented on dipy/core/tests/test_gradients.py:
```diff
@@ -461,6 +460,26 @@ def test_generate_bvecs():
+    assert np.array_equal(gtab_slice1.bvecs, np.array([[1., 0., 0.]]))
+
+    # Test with a range of indices
+    gtab_slice2 = gtab.__get_item__(slice(2, 5))
```
same, the goal is to do `gtab[2:5]` and not `gtab.__get_item__`.. Please can you update

shilpiprd replied:
Sure

shilpiprd replied:
It seems there was a typo in the test file. I changed it, it should pass all the tests now.

skoudoro reviewed (commented):
Hi @shilpiprd,

Can you address my comments below? Thank you

shilpiprd commented:
I don't think it's my method that's failing. It says 1 check failing, but inside it says PASSED for all.

skoudoro commented:
Thank you @shilpiprd, great job for your first contribution! it works like a charm. merging.

jhlegarreta commented:
I was debugging some code and had a look at this line:
https://github.com/dipy/dipy/commit/3fae50e79d6830627633ba84839499134ad63c6b#diff-bbd3ad0756880932ac2142d609f873674431fa6ad7a187ae379bba84ab2ec06fR225

This certainly looks wrong (or weird to say the least): `self.b0_threshold` is a b-value, and `slice_start` is an index, so it makes no sense to assign an index value to `self.b0_threshold`.

The comparison of the `slice_start` variable (an index) to the b0 threshold does not look right as well.

I have gone through this:
https://github.com/dipy/dipy/pull/2749#discussion_r1153704938

And it is unclear to me why the b0 threshold value should be updated if we are just doing a slicing operation (at least in the most general case of the slicing operation).

Should I be wrong and the implementation be OK, @shilpiprd @skoudoro can this be explained in a comment in the code, please?

skoudoro commented:
> This certainly looks wrong (or weird to say the least): self.b0_threshold is a b-value, and slice_start is an index, so it makes no sense to assign an index value to self.b0_threshold.

yep, looks wrong to me too after a quick look. I suppose the idea was 
`self.self.b0_threshold = self.b0_threshold[slice_start]`. but I do not remember the context.  I am not even sure if doing that make sense.

I will go through the discussion later tonight to refresh my memory

shilpiprd commented:
@jhlegarreta , Actually, I think you might be right. I'm not sure of the context since I didn't write that code. But at the first glance, it does seem like index assignment is wrong.
